File: C++/016_3Sum Closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int> &num, int target) {
        if (num.size() < 3)
            return -1;
        sort(num.begin(), num.end());
        int res = num[0] + num[1] + num[2];
        for (int i = 0; i < num.size(); i++) {
            int j = i + 1, k = (int)num.size() - 1;
            int tar = target - num[i];
            while (j < k) {
                if (abs(tar - num[j] - num[k]) < abs(target - res))
                    res = num[i] + num[j] + num[k];
                if (num[j] + num[k] == tar)
                    return res;
                else if (num[j] + num[k] < tar)
                    j++;
                else
                    k--;
            }
        }
        return res;
    }
};

File: Python/016_3Sum Closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if nums is None or len(nums) < 3:
            return 0
        nums.sort()
        res = nums[0] + nums[1] + nums[2]
        for i in range(len(nums)):
            j, k = i + 1, len(nums) - 1
            while j < k:
                t = nums[i] + nums[j] + nums[k]
                if t == target:
                    return target
                res = t if abs(target - t) < abs(target - res) else res
                if t < target:
                    j += 1
                else:
                    k -= 1
        return res